File: internal/model/alerts_test.go
package model

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestAlertDefaults(t *testing.T) {
	var s AlertSettings // settings saved before alerts existed
	s.ApplyDefaults()
	if s.Enabled || len(s.SiteRules) == 0 || len(s.ServerRules) == 0 || s.RecoveryMinutes != 2 || s.EmailTo == nil {
		t.Fatalf("defaults = %+v", s)
	}
	if err := s.Validate(); err != nil {
		t.Fatal(err)
	}
	// A saved configuration keeps its choices: rules removed stay removed.
	s = AlertSettings{Enabled: true, SiteRules: []AlertRule{}}
	s.ApplyDefaults()
	if len(s.SiteRules) != 0 || len(s.ServerRules) != 0 || s.RecoveryMinutes != 2 {
		t.Fatalf("partial = %+v", s)
	}
	// Unnamed rules are named after their metric; windowed ones get a window.
	s = AlertSettings{Enabled: true, SiteRules: []AlertRule{
		{Metric: AlertErrorRate, Threshold: 5}, {Metric: AlertErrorRate, Threshold: 10}, {ID: "cpu", Metric: AlertCPU, Threshold: 80, WindowMinutes: 9},
	}}
	s.ApplyDefaults()
	r := s.SiteRules
	if r[0].ID != "errorrate" || r[1].ID != "errorrate-2" || r[0].WindowMinutes != 5 || r[0].MinRequests != 20 || r[0].Severity != SeverityWarning {
		t.Fatalf("named = %+v", r)
	}
	if r[2].WindowMinutes != 0 {
		t.Fatalf("a window on a non-rate metric was kept: %+v", r[2])
	}
	if err := s.Validate(); err != nil {
		t.Fatal(err)
	}
}

func TestAlertValidation(t *testing.T) {
	cases := map[string]func(*AlertSettings){
		"alerts.siteRules[0].metric":        func(s *AlertSettings) { s.SiteRules[0].Metric = "load" },
		"alerts.siteRules[1].metric":        func(s *AlertSettings) { s.SiteRules[1].Metric = AlertDiskFree },
		"alerts.serverRules[0].metric":      func(s *AlertSettings) { s.ServerRules[0].Metric = AlertCPU },
		"alerts.siteRules[0].threshold":     func(s *AlertSettings) { s.SiteRules[0].Threshold = -1 },
		"alerts.serverRules[2].threshold":   func(s *AlertSettings) { s.ServerRules[2].Threshold = 101 },
		"alerts.siteRules[0].forMinutes":    func(s *AlertSettings) { s.SiteRules[0].ForMinutes = 2000 },
		"alerts.siteRules[0].severity":      func(s *AlertSettings) { s.SiteRules[0].Severity = "page" },
		"alerts.siteRules[3].windowMinutes": func(s *AlertSettings) { s.SiteRules[3].WindowMinutes = 90 },
		"alerts.siteRules[0].repeatHours":   func(s *AlertSettings) { s.SiteRules[0].RepeatHours = 1000 },
		"alerts.siteRules[1].id":            func(s *AlertSettings) { s.SiteRules[1].ID = "CPU" },
		"alerts.serverRules[0].id":          func(s *AlertSettings) { s.ServerRules[0].ID = "errors" },
		"alerts.siteRules[2].id":            func(s *AlertSettings) { s.SiteRules[2].ID = "has space" },
		"alerts.recoveryMinutes":            func(s *AlertSettings) { s.RecoveryMinutes = 61 },
		"alerts.emailTo[1]":                 func(s *AlertSettings) { s.EmailTo = []string{"ops@example.com", "ops"} },
	}
	for field, mutate := range cases {
		s := DefaultAlerts()
		mutate(&s)
		s.ApplyDefaults()
		err := s.Validate()
		ve, ok := err.(*ValidationError)
		if !ok || ve.Field != field {
			t.Errorf("%s: got %v", field, err)
		}
	}
}

func TestSiteAlertValidation(t *testing.T) {
	static := &Site{Name: "files", Type: SiteStatic, Static: &StaticConfig{Root: `C:\www`}}
	static.Alerts.Rules = []AlertRule{{Metric: AlertEventLoopLag, Threshold: 100}}
	static.Alerts.applyDefaults()
	if static.Alerts.Rules[0].ID != "site-eventlooplag" {
		t.Fatalf("id = %q", static.Alerts.Rules[0].ID)
	}
	err := static.Alerts.validate(static)
	if ve, ok := err.(*ValidationError); !ok || ve.Field != "alerts.rules[0].metric" || !strings.Contains(ve.Message, "Node.js") {
		t.Fatalf("lag on a static site: %v", err)
	}
	// Turning a server-wide rule off is fine whatever its metric.
	static.Alerts.Rules[0].Disabled = true
	if err := static.Alerts.validate(static); err != nil {
		t.Fatal(err)
	}

	worker := &Site{Name: "queue", Type: SiteWorker, Node: &NodeConfig{}}
	worker.Alerts.Rules = []AlertRule{{ID: "e", Metric: AlertErrorRate, Threshold: 5, Severity: SeverityCritical, WindowMinutes: 5}}
	if err := worker.Alerts.validate(worker); err == nil || !strings.Contains(err.Error(), "workers") {
		t.Fatalf("error rate on a worker: %v", err)
	}
	worker.Alerts.Rules[0] = AlertRule{ID: "disk", Metric: AlertDiskFree, Threshold: 5, Severity: SeverityWarning}
	if err := worker.Alerts.validate(worker); err == nil || !strings.Contains(err.Error(), "server metric") {
		t.Fatalf("server metric on a site: %v", err)
	}
}

func TestEffectiveAlertRules(t *testing.T) {
	defaults := DefaultAlerts().SiteRules
	node := &Site{Name: "api", Type: SiteNode, Node: &NodeConfig{Instances: 2}}
	ids := func(rules []AlertRule) string {
		var out []string
		for _, r := range rules {
			out = append(out, r.ID)
		}
		return strings.Join(out, ",")
	}
	// No memory limit and no agent: those rules cannot be measured.
	if got := ids(EffectiveAlertRules(defaults, node)); got != "cpu,errors,latency-p95,instances-down" {
		t.Fatalf("node = %s", got)
	}
	node.Node.AgentEnabled = true
	node.Node.Recycle.MemoryLimitMB = 512
	if got := ids(EffectiveAlertRules(defaults, node)); got != "cpu,memory-limit,event-loop-lag,errors,latency-p95,instances-down" {
		t.Fatalf("node with agent and limit = %s", got)
	}
	// Overrides replace in place, off turns off, own rules follow.
	node.Alerts.Rules = []AlertRule{
		{ID: "site-memory", Metric: AlertMemory, Threshold: 800},
		{ID: "ERRORS", Metric: AlertErrorRate, Threshold: 1},
		{ID: "latency-p95", Metric: AlertLatencyP95, Disabled: true},
	}
	eff := EffectiveAlertRules(defaults, node)
	if got := ids(eff); got != "cpu,memory-limit,event-loop-lag,ERRORS,instances-down,site-memory" {
		t.Fatalf("overridden = %s", got)
	}
	if eff[3].Threshold != 1 {
		t.Fatalf("override = %+v", eff[3])
	}
	node.Alerts.Disabled = true
	if len(EffectiveAlertRules(defaults, node)) != 0 {
		t.Fatal("an opted-out site has rules")
	}

	static := &Site{Name: "files", Type: SiteStatic}
	if got := ids(EffectiveAlertRules(defaults, static)); got != "errors,latency-p95" {
		t.Fatalf("static = %s", got)
	}
	worker := &Site{Name: "queue", Type: SiteWorker, Node: &NodeConfig{}}
	if got := ids(EffectiveAlertRules(defaults, worker)); got != "cpu,instances-down" {
		t.Fatalf("worker = %s", got)
	}
}

func TestMemoryLimit(t *testing.T) {
	n := NodeConfig{}
	if n.MemoryLimitMB() != 0 {
		t.Fatal("no limit")
	}
	n.Recycle.MemoryLimitMB = 1024
	n.Limits.MemoryLimitMB = 2048
	if n.MemoryLimitMB() != 1024 {
		t.Fatalf("limit = %d", n.MemoryLimitMB())
	}
	n.Recycle.MemoryLimitMB = 0
	if n.MemoryLimitMB() != 2048 {
		t.Fatalf("limit = %d", n.MemoryLimitMB())
	}
}

// A site saved before alerts existed decodes with no alert configuration
// and stays valid.
func TestSiteWithoutAlerts(t *testing.T) {
	var s Site
	if err := json.Unmarshal([]byte(`{"name":"old","type":"redirect","redirect":{"targetUrl":"https://example.com","statusCode":301}}`), &s); err != nil {
		t.Fatal(err)
	}
	s.ApplyDefaults()
	if s.Alerts.Disabled || len(s.Alerts.Rules) != 0 {
		t.Fatalf("alerts = %+v", s.Alerts)
	}
	if err := s.Validate(); err != nil {
		t.Fatal(err)
	}
}

File: web/src/lib/alerts.test.ts
import { describe, expect, it } from 'vitest';
import type { AlertRule } from '@/api/alerts';
import type { Site } from '@/api/types';
import {
  ALERT_METRICS,
  alertCounts,
  alertSubject,
  defaultAlerts,
  describeRule,
  effectiveRules,
  formatAlertValue,
  memoryLimitMB,
  newRule,
  normalizeAlerts,
  normalizeRule,
  ruleApplies,
  ruleError,
  silenceText,
  withSiteRule,
  withoutSiteRule,
} from './alerts';
import { defaultNode, newSite } from './siteDefaults';

function nodeSite(p: { agent?: boolean; limit?: number } = {}): Site {
  const s = newSite('node');
  s.node = { ...defaultNode(), agentEnabled: !!p.agent };
  s.node.recycle = { ...s.node.recycle, memoryLimitMB: p.limit };
  return s;
}

describe('normalizeAlerts', () => {
  it('fills settings saved before alerts existed', () => {
    expect(normalizeAlerts(undefined)).toEqual(defaultAlerts());
    expect(normalizeAlerts({} as never)).toEqual(defaultAlerts());
  });
  it('keeps a saved configuration', () => {
    const a = normalizeAlerts({ enabled: true, siteRules: [], serverRules: null, recoveryMinutes: 0, emailTo: null });
    expect(a).toEqual({ enabled: true, siteRules: [], serverRules: [], recoveryMinutes: 2, emailTo: [] });
  });
  it('fills rule defaults like the server', () => {
    expect(normalizeRule({ id: 'e', metric: 'errorRate', threshold: 5, forMinutes: 5, severity: '' })).toEqual({
      id: 'e', metric: 'errorRate', threshold: 5, forMinutes: 5, severity: 'warning', windowMinutes: 5, minRequests: 20,
    });
    expect(normalizeRule({ id: 'c', metric: 'cpu', threshold: 5, forMinutes: 5, severity: 'critical', windowMinutes: 9 })).not.toHaveProperty('windowMinutes');
  });
});

describe('newRule', () => {
  it('picks a free ID', () => {
    expect(newRule('cpu', []).id).toBe('cpu');
    expect(newRule('cpu', ['CPU', 'cpu-2']).id).toBe('cpu-3');
    expect(newRule('latencyP95', ['latency-p95'], 'site-')).toMatchObject({ id: 'site-latencyp95', threshold: 2000, windowMinutes: 5 });
  });
});

describe('ruleApplies', () => {
  it('matches the site type and its configuration', () => {
    const node = nodeSite();
    expect(ruleApplies({ metric: 'instanceCpu' }, node)).toBe(true);
    expect(ruleApplies({ metric: 'memoryPercent' }, node)).toBe(false);
    expect(ruleApplies({ metric: 'memoryPercent' }, nodeSite({ limit: 512 }))).toBe(true);
    expect(ruleApplies({ metric: 'eventLoopLag' }, node)).toBe(false);
    expect(ruleApplies({ metric: 'eventLoopLag' }, nodeSite({ agent: true }))).toBe(true);
    expect(ruleApplies({ metric: 'diskFree' }, node)).toBe(false);
    const stat = newSite('static');
    expect(ruleApplies({ metric: 'errorRate' }, stat)).toBe(true);
    expect(ruleApplies({ metric: 'cpu' }, stat)).toBe(false);
    const worker = newSite('worker');
    expect(ruleApplies({ metric: 'errorRate' }, worker)).toBe(false);
    expect(ruleApplies({ metric: 'instancesDown' }, worker)).toBe(true);
  });
  it('uses the lower memory limit', () => {
    const s = nodeSite({ limit: 1024 });
    s.node!.limits = { ...s.node!.limits, memoryLimitMB: 800 };
    expect(memoryLimitMB(s)).toBe(800);
    expect(memoryLimitMB(newSite('static'))).toBe(0);
  });
});

describe('effectiveRules', () => {
  const defaults = defaultAlerts().siteRules!;
  it('lists server rules, overrides in place and site rules after', () => {
    const s = nodeSite({ agent: true });
    s.alerts = {
      rules: [
        { id: 'site-memory', metric: 'memory', threshold: 800, forMinutes: 5, severity: 'warning' },
        { id: 'ERRORS', metric: 'errorRate', threshold: 1, forMinutes: 5, severity: 'critical', windowMinutes: 5, minRequests: 20 },
        { id: 'latency-p95', metric: 'latencyP95', threshold: 2000, forMinutes: 10, severity: 'warning', disabled: true },
      ],
    };
    const eff = effectiveRules(defaults, s);
    expect(eff.map((e) => `${e.rule.id}:${e.source}${e.off ? ':off' : ''}${e.skipped ? ':skip' : ''}`)).toEqual([
      'cpu:server',
      'memory-limit:server:skip',
      'event-loop-lag:server',
      'ERRORS:override',
      'latency-p95:override:off',
      'instances-down:server',
      'site-memory:site',
    ]);
    expect(eff[3].base?.threshold).toBe(5);
    expect(eff[1].skipped).toBe('no memory limit set');
  });
  it('says why rules do not apply', () => {
    const eff = effectiveRules(defaults, newSite('static'));
    expect(eff.filter((e) => !e.skipped).map((e) => e.rule.id)).toEqual(['errors', 'latency-p95']);
    expect(eff[0].skipped).toBe('Node.js sites only');
  });
});

describe('withSiteRule / withoutSiteRule', () => {
  const r: AlertRule = { id: 'cpu', metric: 'instanceCpu', threshold: 95, forMinutes: 10, severity: 'warning' };
  it('adds, replaces and removes', () => {
    let a = withSiteRule(undefined, r);
    expect(a.rules).toEqual([r]);
    a = withSiteRule({ disabled: false, ...a }, { ...r, id: 'CPU', threshold: 99 });
    expect(a.rules).toHaveLength(1);
    expect(a.rules![0].threshold).toBe(99);
    a = withoutSiteRule(a, 'cpu');
    expect(a).toEqual({ disabled: false });
  });
});

describe('formatting', () => {
  it('formats values like the server', () => {
    expect(formatAlertValue('cpu', 93.4)).toBe('93%');
    expect(formatAlertValue('errorRate', 2.46)).toBe('2.5%');
    expect(formatAlertValue('memory', 1234.4)).toBe('1,234 MB');
    expect(formatAlertValue('latencyP95', 850.2)).toBe('850 ms');
    expect(formatAlertValue('latencyP95', 2430)).toBe('2.4 s');
    expect(formatAlertValue('instancesDown', 2)).toBe('2');
  });
  it('describes rules', () => {
    const d = defaultAlerts();
    expect(describeRule(d.siteRules![0])).toBe('CPU of the busiest instance above 90% for 10 min');
    expect(describeRule(d.siteRules![3])).toBe('5xx error rate above 5% (over 5 min, at least 20 requests) for 5 min');
    expect(describeRule(d.siteRules![5])).toBe('Any instance down for 5 min');
    expect(describeRule({ ...d.siteRules![5], threshold: 2, forMinutes: 0 })).toBe('More than 2 instances down');
    expect(describeRule(d.serverRules![2])).toBe('Free disk space below 10% for 5 min');
  });
  it('describes silences and subjects', () => {
    const t = (iso: string) => iso.slice(11, 16);
    expect(silenceText({}, t)).toBe('');
    expect(silenceText({ silence: { by: 'bob', at: '' } }, t)).toBe('acknowledged by bob');
    expect(silenceText({ silence: { by: 'al', at: '', until: '2026-09-24T14:30:00Z' } }, t)).toBe('until 14:30 by al');
    expect(alertSubject({ siteId: 's', siteName: 'api' })).toBe('api');
    expect(alertSubject({})).toBe('Server');
  });
});

describe('ruleError', () => {
  const ok: AlertRule = { id: 'x', metric: 'errorRate', threshold: 5, forMinutes: 5, severity: 'warning', windowMinutes: 5, minRequests: 20 };
  it('accepts the defaults', () => {
    for (const r of [...defaultAlerts().siteRules!, ...defaultAlerts().serverRules!]) expect(ruleError(r)).toBeNull();
  });
  it.each<[Partial<AlertRule>, string]>([
    [{ metric: 'load' }, 'metric'],
    [{ id: 'has space' }, 'ID'],
    [{ threshold: 101 }, 'limit'],
    [{ forMinutes: 2000 }, 'period'],
    [{ windowMinutes: 45 }, 'window'],
    [{ repeatHours: 200 }, 'Reminders'],
  ])('rejects %j', (patch, word) => {
    expect(ruleError({ ...ok, ...patch })).toContain(word);
  });
});

describe('alertCounts', () => {
  it('counts firing alerts by severity, silenced apart', () => {
    const a = (severity: string, silenced = false) =>
      ({ id: '', ruleId: '', metric: 'cpu', severity, threshold: 0, forMinutes: 0, state: 'firing', value: 0, peak: 0, message: '', since: '', notified: true, silence: silenced ? { by: 'x', at: '' } : undefined });
    expect(alertCounts({ enabled: true, firing: [a('critical'), a('warning'), a('critical', true)], pending: [a('critical')] })).toEqual({ critical: 1, warning: 1, silenced: 1 });
    expect(alertCounts(undefined)).toEqual({ critical: 0, warning: 0, silenced: 0 });
  });
});

it('has a unique key per metric', () => {
  expect(new Set(ALERT_METRICS.map((m) => m.key)).size).toBe(ALERT_METRICS.length);
});
